### PR TITLE
expose experimental option to set a memory limit for turbo

### PR DESCRIPTION
**Pull request by arlyon**
### What?

We have a GC with rough edges, but it is worth exposing a flag to be able to try it out.

### Why?

I think it would be useful to set up a benchmark that tests the GC perf on new canary builds to make sure we are going in the right direction.

### How?

Add a new flag in the experimental settings and wire it up to the existing native bindings.

Closes PACK-3093

### CODE REVIEW

**Comment by arlyon** [comment]
* **#66101** <a href="https://app.graphite.dev/github/pr/vercel/next.js/66101?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arlyon and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: aea323c411617a8814d9235b755ac76dccfef50d

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: aea323c411617a8814d9235b755ac76dccfef50d

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

File: packages/next/src/build/index.ts [hunk 1]
```diff
@@ -1378,29 +1378,34 @@ export default async function build(
...
-          nextConfig: config,
-          jsConfig: await getTurbopackJsConfig(dir, config),
-          watch: false,
-          dev,
-          env: process.env as Record<string, string>,
-          defineEnv: createDefineEnv({
-            isTurbopack: true,
-            clientRouterFilters: NextBuildContext.clientRouterFilters,
-            config,
+        const project = await bindings.turbo.createProject(
```
Comment (arlyon) [inline packages/next/src/build/index.ts:1381]: Looks like husky formatted this automatically?

File: packages/next/src/server/dev/hot-reloader-turbopack.ts [hunk 2]
```diff
@@ -123,30 +123,35 @@ export async function createHotReloaderTurbopack(
...
-    jsConfig: await getTurbopackJsConfig(dir, nextConfig),
-    watch: true,
-    dev: true,
-    env: process.env as Record<string, string>,
-    defineEnv: createDefineEnv({
-      isTurbopack: true,
-      // TODO: Implement
-      clientRouterFilters: undefined,
-      config: nextConfig,
+  const project = await bindings.turbo.createProject(
```
Comment (arlyon) [inline packages/next/src/server/dev/hot-reloader-turbopack.ts:126]: Same again...

**Comment by ijjk** [comment]
## Failing test suites 
Commit: aea323c411617a8814d9235b755ac76dccfef50d

`pnpm test-dev test/e2e/app-dir/navigation/navigation.test.ts`
- app dir - navigation > navigating to dynamic params & changing the casing > should load the page correctly
- app dir - navigation > query string > should set query correctly

<details>
<summary>Expand output</summary>

  ● app dir - navigation › query string › should set query correctly

    page.goto: Timeout 60000ms exceeded.
    Call log:
      - navigating to "http://localhost:43857/", waiting until "load"

      274 |     opts?.beforePageLoad?.(page)
      275 |
    > 276 |     await page.goto(url, { waitUntil: 'load' })
          |                ^
      277 |   }
      278 |
      279 |   back(options): BrowserInterface {

      at BrowserInterface.goto (lib/browsers/playwright.ts:276:16)
      at webdriver (lib/next-webdriver.ts:129:3)
      at Object.<anonymous> (e2e/app-dir/navigation/navigation.test.ts:12:23)

  ● app dir - navigation › navigating to dynamic params & changing the casing › should load the page correctly

    expect(received).toContain(expected) // indexOf

    Expected substring: "[paramB] page"
    Received string:    "/paramA/paramB
    /paramA/noParam"

      881 |
      882 |       await retry(async () => {
    > 883 |         expect(await browser.elementByCss('body').text()).toContain(
          |                                                           ^
      884 |           '[paramB] page'
      885 |         )
      886 |       })

      at toContain (e2e/app-dir/navigation/navigation.test.ts:883:59)
      at retry (lib/next-test-utils.ts:774:14)
      at Object.<anonymous> (e2e/app-dir/navigation/navigation.test.ts:882:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| buildDuration | 17.3s | 17.2s | N/A |
| buildDurationCached | 8.1s | 7s | N/A |
| nodeModulesSize | 351 MB | 351 MB | N/A |
| nextStartRea..uration (ms) | 417ms | 418ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| 3430.HASH.js gzip | 169 B | 169 B | ✓ |
| 3601-HASH.js gzip | 5.05 kB | 5.06 kB | N/A |
| 5450-HASH.js gzip | 35.5 kB | 35.2 kB | N/A |
| 67a0f7eb-HASH.js gzip | 51.4 kB | 51.5 kB | N/A |
| framework-HASH.js gzip | 56.5 kB | 56.5 kB | N/A |
| main-app-HASH.js gzip | 221 B | 216 B | N/A |
| main-HASH.js gzip | 32.3 kB | 32.3 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 169 B | 169 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 194 B | N/A |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 512 B | 511 B | N/A |
| css-HASH.js gzip | 343 B | 341 B | N/A |
| dynamic-HASH.js gzip | 2.52 kB | 2.53 kB | N/A |
| edge-ssr-HASH.js gzip | 266 B | 265 B | N/A |
| head-HASH.js gzip | 363 B | 364 B | N/A |
| hooks-HASH.js gzip | 392 B | 390 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 327 B | 329 B | N/A |
| script-HASH.js gzip | 398 B | 398 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 324 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 964 B | 964 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| _buildManifest.js gzip | 482 B | 483 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| index.html gzip | 522 B | 519 B | N/A |
| link.html gzip | 536 B | 535 B | N/A |
| withRouter.html gzip | 518 B | 516 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| edge-ssr.js gzip | 124 kB | 124 kB | N/A |
| page.js gzip | 188 kB | 188 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 658 B | 654 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 29.6 kB | 29.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.02 kB | 1.02 kB | ✓ |
| Overall change | 1.02 kB | 1.02 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 185 kB | 185 kB | N/A |
| app-page-exp..prod.js gzip | 112 kB | 112 kB | N/A |
| app-page-tur..prod.js gzip | 123 kB | 123 kB | N/A |
| app-page-tur..prod.js gzip | 119 kB | 119 kB | N/A |
| app-page.run...dev.js gzip | 177 kB | 177 kB | N/A |
| app-page.run..prod.js gzip | 109 kB | 109 kB | N/A |
| app-route-ex...dev.js gzip | 24.5 kB | 24.5 kB | N/A |
| app-route-ex..prod.js gzip | 18.3 kB | 18.3 kB | N/A |
| app-route-tu..prod.js gzip | 18.3 kB | 18.3 kB | N/A |
| app-route-tu..prod.js gzip | 18.1 kB | 18.1 kB | N/A |
| app-route.ru...dev.js gzip | 24.2 kB | 24.2 kB | N/A |
| app-route.ru..prod.js gzip | 18.1 kB | 18.1 kB | N/A |
| pages-api-tu..prod.js gzip | 9.53 kB | 9.53 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.53 kB | 9.53 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| pages.runtim...dev.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim..prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| server.runti..prod.js gzip | 52.1 kB | 52 kB | N/A |
| Overall change | 93.7 kB | 93.7 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js arlyon/expose-memory-limit | Change |
| - | - | - | - |
| 0.pack gzip | 1.67 MB | 1.66 MB | N/A |
| index.pack gzip | 129 kB | 128 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1122: /***/ (
+    /***/ 2058: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(3294);
+          return __webpack_require__(6302);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 2985: /***/ (module, exports, __webpack_require__) => {
+    /***/ 6392: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(6024)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(7364)
+        __webpack_require__(7174)
       );
-      const _getimgprops = __webpack_require__(3108);
-      const _imageconfig = __webpack_require__(9828);
-      const _imageconfigcontextsharedruntime = __webpack_require__(9053);
-      const _warnonce = __webpack_require__(587);
-      const _routercontextsharedruntime = __webpack_require__(3766);
+      const _getimgprops = __webpack_require__(4804);
+      const _imageconfig = __webpack_require__(2977);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9777);
+      const _warnonce = __webpack_require__(3257);
+      const _routercontextsharedruntime = __webpack_require__(3543);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(3189)
+        __webpack_require__(8406)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -376,7 +376,7 @@
       /***/
     },
 
-    /***/ 3108: /***/ (
+    /***/ 4804: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -392,9 +392,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(587);
-      const _imageblursvg = __webpack_require__(4779);
-      const _imageconfig = __webpack_require__(9828);
+      const _warnonce = __webpack_require__(3257);
+      const _imageblursvg = __webpack_require__(8719);
+      const _imageconfig = __webpack_require__(2977);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -769,7 +769,7 @@
       /***/
     },
 
-    /***/ 4779: /***/ (__unused_webpack_module, exports) => {
+    /***/ 8719: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -824,7 +824,7 @@
       /***/
     },
 
-    /***/ 6660: /***/ (
+    /***/ 260: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -851,10 +851,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(1478);
-      const _getimgprops = __webpack_require__(3108);
-      const _imagecomponent = __webpack_require__(2985);
+      const _getimgprops = __webpack_require__(4804);
+      const _imagecomponent = __webpack_require__(6392);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(3189)
+        __webpack_require__(8406)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -886,7 +886,7 @@
       /***/
     },
 
-    /***/ 3189: /***/ (__unused_webpack_module, exports) => {
+    /***/ 8406: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -921,7 +921,7 @@
       /***/
     },
 
-    /***/ 3294: /***/ (
+    /***/ 6302: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -938,8 +938,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@19.0.0-rc-81c5ff2e04-20240521/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(5293);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-81c5ff2e04-20240521_react_52riaj2pppk6biljiae3nnywwa/node_modules/next/image.js
-      var next_image = __webpack_require__(6828);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-81c5ff2e04-20240521_react_qjsttk5yohmngz5jvpmxy4vvtq/node_modules/next/image.js
+      var next_image = __webpack_require__(3132);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -969,12 +969,12 @@
       /***/
     },
 
-    /***/ 6828: /***/ (
+    /***/ 3132: /***/ (
       module,
       __unused_webpack_exports,
       __webpack_require__
     ) => {
-      module.exports = __webpack_require__(6660);
+      module.exports = __webpack_require__(260);
 
       /***/
     },
@@ -984,7 +984,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1122)
+      __webpack_exec__(2058)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>5450-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: aea323c411617a8814d9235b755ac76dccfef50d

**Review by bgw — APPROVED** [review]

File: packages/next/src/server/config-shared.ts [hunk 3]
```diff
@@ -195,6 +195,10 @@ export interface ReactCompilerOptions {
 }
 
 export interface ExperimentalConfig {
+  /**
+   * A target memory limit for turbo, in bytes.
+   */
+  turboMemoryLimit?: number
```
Comment (ijjk) [inline packages/next/src/server/config-shared.ts:201]: This needs to be added to the config-schema as well https://github.com/vercel/next.js/blob/canary/packages/next/src/server/config-schema.ts and we should set a default value (even if `undefined`) in this file.

**Pull request merged**